arbutil/block_message_relation.go

```diff
@@ -3,17 +3,13 @@
 
 package arbutil
 
+// messages are 0-indexed
 type MessageIndex uint64
 
 func BlockNumberToMessageCount(blockNumber uint64, genesisBlockNumber uint64) MessageIndex {
 	return MessageIndex(blockNumber + 1 - genesisBlockNumber)
 }
```

Comment:
What do you think about having a single type that represents the position which is what you pass around, and then having accessors for `.Index()` and `.Count()`?

You could even wrap it like this so you don't accidentally do some illegal operation with it.
```
type MessageIndex struct {
  inner uint64
}
```

I'm interested to hear others' thoughts too on this part of the PR since it's the core of the design.

Reply:
But what is the meaning of `.Count()`?
It is the current length of the messages container, the expected length of the messages container, or just `.Index() + 1`?